# WIB_scRNA-seq

This repository contains the materials for the **Boston-area Women's Bioinformatics Meetup** for *Cluster without the fluster: what to expect in a single-cell RNA-seq workflow*. The materials highlight the different steps and considerations involved when performing a single-cell RNA-seq analysis, from raw data QC to clustering and marker identification.

**Materials:**

- [Raw data to generation of the count matrix](https://github.com/marypiper/WIB_scRNA-seq/blob/master/lessons/SC_pre-QC.md)
- [Quality control and filtering](https://github.com/marypiper/WIB_scRNA-seq/blob/master/lessons/SC_quality_control_analysis.md)
- [Integration of datasets and clustering](https://github.com/marypiper/WIB_scRNA-seq/blob/master/lessons/SC_clustering_analysis.md)
- [Marker identification and identification of cell types](https://github.com/marypiper/WIB_scRNA-seq/blob/master/lessons/SC_marker_identification.md)


***

*These materials have been developed by Mary Piper. These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*

* *These materials were derived from work that is Copyright Â© [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). 
All HBC instructional material is made available under the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0).*
* *Adapted from the lessons by [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/) with contributions from Mary Piper, Radhika Khetani, Meeta Mistry, Rory Kirchner, and Lorena Pantano.*
* *A portion of these materials and hands-on activities were adapted from the [Satija Lab's](https://satijalab.org/) [Seurat - Guided Clustering Tutorial](https://satijalab.org/seurat/pbmc3k_tutorial.html)*
